avogadro/qtplugins/labels/labels.cpp

```diff
@@ -0,0 +1,188 @@
+#include <avogadro/rendering/cylindergeometry.h>
+#include <avogadro/rendering/geometrynode.h>
+#include <avogadro/rendering/groupnode.h>
+#include <avogadro/rendering/spheregeometry.h>
+#include <avogadro/rendering/textlabel3d.h>
+#include <avogadro/rendering/textproperties.h>
+
+#include <QtWidgets/QCheckBox>
+#include <QtWidgets/QDoubleSpinBox>
+#include <QtWidgets/QHBoxLayout>
+#include <QtWidgets/QLabel>
+#include <QtWidgets/QVBoxLayout>
+#include <QtWidgets/QWidget>
+
+namespace Avogadro {
+namespace QtPlugins {
+
+using Avogadro::QtGui::RWAtom;
+using Avogadro::QtGui::RWBond;
+using Avogadro::QtGui::RWMolecule;
+using Core::Elements;
+using Core::Molecule;
+using Rendering::GeometryNode;
+using Rendering::GroupNode;
+using Rendering::SphereGeometry;
+using Rendering::CylinderGeometry;
+
+typedef Avogadro::Core::Array<Molecule::BondType> NeighborListType;
+
+using namespace std;
+
+Labels::Labels(QObject* p)
+  : ScenePlugin(p), m_enabled(false), m_group(nullptr), m_setupWidget(nullptr),
+    m_multiBonds(true), m_showHydrogens(true)
+{
+}
+
+Labels::~Labels()
+{
+  if (m_setupWidget)
+    m_setupWidget->deleteLater();
+}
+
+void Labels::process(const Molecule& molecule, Rendering::GroupNode& node)
+{
+  Index hydrogenCount = 1;
+
+  m_group = &node;
+  GeometryNode* geometry = new GeometryNode;
+  node.addChild(geometry);
+
+  std::unordered_map<unsigned char, Index>
+    countMap; // Dictionary of atomic number and number of atoms
+
+  for (Index i = 0; i < molecule.atomCount(); ++i) {
+    Core::Atom atom = molecule.atom(i);
+    unsigned char atomicNumber = atom.atomicNumber();
+    if (atomicNumber == 1)
+      continue; // Do not process Hydrogens separately
+
+    std::unordered_map<unsigned char, Index>::const_iterator lookUp;
+    lookUp = countMap.find(atomicNumber);
+
+    if (lookUp == countMap.end())
+      countMap[atomicNumber] = 1; // For first occurrence of atom
+
+    else
+      countMap[atomicNumber] += 1; // For other occurrences of atom
+
+    if (!m_showHydrogens)
+      continue;
+
+    Rendering::TextLabel3D* atomLabel = new Rendering::TextLabel3D;
+
+    //  Label non Hydrogen atoms
+    atomLabel->setText(
+      Elements::symbol(atomicNumber) +
+      QString::number(countMap[atomicNumber], 'f', 0).toStdString());
+
+    atomLabel->setRenderPass(Rendering::Overlay3DPass);
+    const Vector3f a1(atom.position3d().cast<float>());
+    const Vector3f& textPos(a1);
+    atomLabel->setAnchor(textPos);
+    geometry->addDrawable(atomLabel);
+
+    Rendering::TextProperties tprop;
+    tprop.setAlign(Rendering::TextProperties::HCenter,
+                   Rendering::TextProperties::VCenter);
+    tprop.setFontFamily(Rendering::TextProperties::SansSerif);
+    tprop.setColorRgb(255, 255, 255);
+    atomLabel->setTextProperties(tprop);
+
+    //  Label Hydrogen atoms
+    const NeighborListType bonds =
+      (const_cast<Molecule&>(molecule))
+        .bonds(atom); // const_cast because bonds() accepts non const objects
```

Comment:
We should avoid this pattern of comments pushing code to another line, prefer to make comments above the code they are referencing, not on the same line, and especially not if that creates multi-line code that could have been kept on less lines otherwise. We should also look at whether we need another signature to avoid the const_cast.